### DISCUSSION TITLE
Need help in 2-D homogeneous flow simulation through Navier-Stokes Modules

### DISCUSSION THREAD

**Discussion by dd19990611** [idaholab/moose#25097] [Q&A Modules: Navier-Stokes]
I want to simulate coolant flow in a 2-D rectangular channel with a homogeneous flow model. But the calculations did not converge when I add a heat flux BCs and the water phase changes. So I'd like to ask the following questions.

1. Is there any case for two-phase flow calculation for reference?

2. Is there anything wrong with my input card?

3. How to improve convergence?

4. Can Water97FluidProperties be used in N-S modules?

**Comment by dd19990611** [comment]
mu = 'mu'
rho = 'rho'
k = 'k'
cp = 'cp'
infinite_porosity= 'infinite_porosity'
velocity_interp_method = 'rc'
advected_interp_method = 'average'
#bcflux = 0.1e5
[GlobalParams]
  rhie_chow_user_object = 'rc'
  heat_flux = 5.75e5
[]

[Functions]
  [bc_flux]
    type = ParsedFunction
    value = 0.1e5*t
  []
[]

[UserObjects]
  [rc]
    type = INSFVRhieChowInterpolator
    u = vel_x
    v = vel_y
    pressure = pressure
  []
  [frac_layer]
    type = LayeredAverage
    variable = frac
    direction = y
    num_layers = 10
  []
[]

[VectorPostprocessors]
  [frac]
    type = SpatialUserObjectVectorPostprocessor
    userobject = frac_layer
    points = '0.0015    0 0
              0.0015 0.01 0
              0.0015 0.02 0
              0.0015 0.03 0
              0.0015 0.04 0
              0.0015 0.05 0
              0.0015 0.06 0
              0.0015 0.07 0
              0.0015 0.08 0
              0.0015 0.09 0
              0.0015 0.10 0'
  []
[]

[Mesh]
  [gen]
    type = GeneratedMeshGenerator
    dim = 2
    xmin = 0
    xmax = 0.003
    ymin = 0
    ymax = 0.1
    nx = 30
    ny = 500
  []
[]

[Variables]
  [vel_x]
    type = INSFVVelocityVariable
    initial_condition = 0
    scaling = 1.0e-3
  []
  [vel_y]
    type = INSFVVelocityVariable
    initial_condition = 0.24
    scaling = 1.0e-3
  []
  [pressure]
    type = INSFVPressureVariable
    initial_condition = 0.1e6
    scaling = 1.0e-5
  []
  [T_fluid]
    type = INSFVEnergyVariable
    initial_condition = 360
    scaling = 1.0e-8
  []
[]

[AuxVariables]
  [U]
    order = CONSTANT
    family = MONOMIAL
    fv = true
  []
  [frac]
    order = CONSTANT
    family = MONOMIAL
    fv = true
  []
[]

[AuxKernels]
  [mag]
    type = VectorMagnitudeAux
    variable = U
    x = vel_x
    y = vel_y
  []
  [frac]
   type = FracAuxkernel
   variable = frac
   infinite_porosity = ${infinite_porosity}
  []
[]

[FVKernels]
  [mass]
    type = INSFVMassAdvection
    variable = pressure
    advected_interp_method = ${advected_interp_method}
    velocity_interp_method = ${velocity_interp_method}
    rho = ${rho}
  []
  [u_advection]
    type = INSFVMomentumAdvection
    variable = vel_x
    velocity_interp_method = ${velocity_interp_method}
    advected_interp_method = ${advected_interp_method}
    rho = ${rho}
    momentum_component = 'x'
  []
  [u_viscosity]
    type = INSFVMomentumDiffusion
    variable = vel_x
    mu = ${mu}
    momentum_component = 'x'
  []
  [u_pressure]
    type = INSFVMomentumPressure
    variable = vel_x
    momentum_component = 'x'
    pressure = pressure
  []
  [u_gravity]
    type = INSFVMomentumGravity
    variable = vel_x
    momentum_component = 'x'
    gravity = '0 -1 0'
    rho = ${rho}
  []
  [u_friction_linear]
    type = INSFVMomentumFriction
    variable = vel_x
    linear_coef_name = friction_coefficient
    momentum_component = 'x'
  []
  [u_friction_quad]
    type = INSFVMomentumFriction
    variable = vel_x
    quadratic_coef_name = friction_coefficientq
    momentum_component = 'x'
  []
  [v_advection]
    type = INSFVMomentumAdvection
    variable = vel_y
    velocity_interp_method = ${velocity_interp_method}
    advected_interp_method = ${advected_interp_method}
    rho = ${rho}
    momentum_component = 'y'
  []
  [v_viscosity]
    type = INSFVMomentumDiffusion
    variable = vel_y
    mu = ${mu}
    momentum_component = 'y'
  []
  [v_pressure]
    type = INSFVMomentumPressure
    variable = vel_y
    momentum_component = 'y'
    pressure = pressure
  []
  [v_gravity]
    type = INSFVMomentumGravity
    variable = vel_y
    momentum_component = 'y'
    gravity = '0 -1 0'
    rho = ${rho}
  []
  [v_friction_linear]
    type = INSFVMomentumFriction
    variable = vel_y
    linear_coef_name = friction_coefficient
    momentum_component = 'y'
  []
  [v_friction_quad]
    type = INSFVMomentumFriction
    variable = vel_y
    quadratic_coef_name = friction_coefficientq
    momentum_component = 'y'
  []
  [temp_conduction]
    type = FVDiffusion
    coeff = 'k'
    variable = T_fluid
  []
  [temp_advection]
    type = INSFVEnergyAdvection
    variable = T_fluid
    velocity_interp_method = ${velocity_interp_method}
    advected_interp_method = ${advected_interp_method}
  []
[]

[FVBCs]
  [no_slip_x]
    type = INSFVNoSlipWallBC
    variable = vel_x
    boundary = 'left right'
    function = '0'
  []
  [inlet_x]
    type = INSFVInletVelocityBC
    variable = vel_x
    boundary = 'bottom'
    function = 0
  []
  [no_slip_y]
    type = INSFVNoSlipWallBC
    variable = vel_y
    boundary = 'left right'
    function = 0
  []
  [inlet_y]
    type = INSFVInletVelocityBC
    variable = vel_y
    boundary = 'bottom'
    function = 0.24
  []
  [inlet_T]
    type = WCNSFVInletTemperatureBC
    variable = T_fluid
    boundary = 'bottom'
    temperature_pp =  360
  []
  [heat_wall]
    type = FVFunctionNeumannBC
    variable = T_fluid
    boundary = 'left right'
    function = bc_flux
  []
  [outlet_P]
    type = INSFVOutletPressureBC
    variable = pressure
    boundary = 'top'
    function = 0.1e6
  []
[]

[FluidProperties]
  [water]
    type = Water97FluidProperties
  []
[]

[Materials]
  #inactive = 'mu cp0 k0'
  [speed_material]
    type = PINSFVSpeedFunctorMaterial
    superficial_vel_x = vel_x
    superficial_vel_y = vel_y
    porosity = 1
  []
  [Re_material]
    type = ReynoldsNumberFunctorMaterial
    speed = speed
    characteristic_length = 0.00545
    rho = ${rho}
    mu = ${mu}
  []
  [friction_coefficient]
    type = ExponentialFrictionMaterial
    friction_factor_name = 'friction_coefficient'
    Re = Re
    c1 = 0.32
    c2 = -0.25
  []
  [friction_coefficientq]
    type = ExponentialFrictionMaterial
    friction_factor_name = 'friction_coefficientq'
    Re = Re
    c1 = 0.64
    c2 = -1.0
  []
  [ins_fv]
    type = INSFVEnthalpyMaterial
    temperature = 'T_fluid'
    rho = ${rho}
  []
  [infinite_porosity]
    type = FracFunctorMaterial
    pressure = pressure
    temperature = T_fluid
    fp = water
    temp = 360
    press = 0.1e6
    vel_x = vel_x
    vel_y = vel_y
    chang = 0.03
    duan = 0.003
    #heat_flux = 5.75e5
  []
  [rho]
    type = RhoPTBFunctorMaterial
    fp = water
    temperature = T_fluid
    pressure = pressure
    infinite_porosity = 0.05 #${infinite_porosity}
  []
  [mu]
    type = MufromPTFunctorMaterial
    fp = water
    temperature = T_fluid
    pressure = pressure
    infinite_porosity = 0.05 #${infinite_porosity}
  []
  [k0]
    type = KfromPTFunctorMaterial
    fp = water
    temperature = T_fluid
    pressure = pressure
    infinite_porosity = 0.05 #${infinite_porosity}
  []
  [cp0]
    type = cpfromPTFunctorMaterial
    fp = water
    temperature = T_fluid
    pressure = pressure
    infinite_porosity = 0.05 #${infinite_porosity}
  []
  [functor_constants]
    type = ADGenericFunctorMaterial
    prop_names = 'k0 cp0'
    prop_values = '${k} ${cp}'
  []
[]

[Executioner]
  type = Transient
  end_time = 10
  dt = 1
  l_max_its = 100
  nl_max_its = 20
  solve_type = PJFNK
  petsc_options = '-snes_ksp_ew'
  petsc_options_iname = '-pc_type -pc_factor_mat_solver_package'
  petsc_options_value = ' lu       superlu_dist'
  nl_abs_tol = 1e-8
  nl_rel_tol = 1e-10
  line_search = none
  l_tol = 8.0e-3
[]

[Outputs]
  exodus = true
  csv = true
[]

**Reply by lindsayad** [reply]
We don't have multi-phase modeling in the modules. See https://github.com/idaholab/moose/issues/22637 for discussion of the issue
